lib/node_modules/@stdlib/lapack/base/dgebal/lib/base.js

```diff
@@ -0,0 +1,351 @@
+			ia2 = offsetA + ( k * strideA2 ); // Follows `k`th column of A
+			ia3 = offsetA + ( k * strideA1 ); // Follows `j`th row of A
+			ia4 = offsetA + ( k * strideA1 ); // Follows `k`th row of A
+
+			for ( j = k; j <= l; j++ ) {
+				canSwap = true;
+				for ( i = k; i <= l; i++ ) {
+					if ( i !== j && A[ offsetA + (i*strideA1) + (j*strideA2) ] !== 0.0 ) {
+						canSwap = false;
+						break;
+					}
+				}
+
+				if ( canSwap ) {
+					scale[ is ] = j;
+					if ( j !== k ) {
+						dswap( l+1, A, strideA1, ia1, A, strideA1, ia2 );
+						dswap( N-k, A, strideA2, ia3, A, strideA2, ia4 );
+					}
+					noconv = true;
+					k += 1;
+
+					is += strideScale;
+					ia2 += strideA2;
+					ia4 += strideA1;
+				}
+
+				ia1 += strideA2;
+				ia3 += strideA1;
+			}
+		}
+	}
+
+	// Initialize scale for non-permuted submatrix
+	is = offsetScale + ( k * strideScale );
+	for ( i = k; i <= l; i++ ) {
+		scale[ is ] = 1.0;
+		is += strideScale;
+	}
+
+	if ( job === 'permute' ) {
+		out[ offsetOut ] = k; // ilo
+		out[ offsetOut + strideOut ] = l; // ihi
+		return 0;
+	}
+
+	// Balance the submatrix in rows K to L, iterative loop for norm reduction (job = 'B')
+	sfmin1 = dlamch( 'S' ) / dlamch( 'P' );
+	sfmax1 = 1.0 / sfmin1;
+	sfmin2 = sfmin1 * sclfac;
+	sfmax2 = 1.0 / sfmin2;
+
+	noconv = true;
+	while ( noconv ) {
+		is = offsetScale + ( k * strideScale ); // Follows scale
+		ia1 = offsetA + ( k * strideA1 ) + ( k * strideA2 ); // Follows A[ k, i ]
+		ia2 = offsetA + ( k * strideA1 ) + ( k * strideA2 ); // Follows A[ i, k ]
+		ia3 = offsetA + ( k * strideA2 ); // follows `i`th column of A
+
+		noconv = false;
+		for ( i = k; i <= l; i++ ) {
+			c = dnrm2( l-k+1, A, strideA1, ia1 );
+			r = dnrm2( l-k+1, A, strideA2, ia2 );
+			ica = idamax( l+1, A, strideA1, ia3 );
+			ca = abs( A[ offsetA + (ica*strideA1) + (i*strideA2) ] );
+			ira = idamax( N-k+1, A, strideA2, ia2 );
+			ra = abs( A[ offsetA + (i*strideA1) + ((ira+k)*strideA2) ] );
+
+			if ( c === 0.0 || r === 0.0 ) {
+				ia1 += strideA2;
+				ia2 += strideA1;
+				is += strideScale;
+				ia3 += strideA2;
+				continue;
+			}
+
+			if ( isnan( c ) || isnan( r ) || isnan( ca ) || isnan( ra ) ) {
+				throw new RangeError( 'should not return NaN' );
+			}
+
+			g = r / sclfac;
+			f = 1.0;
+			s = c + r;
+
+			while ( c < g && max( f, c, ca ) < sfmax2 && min( r, g, ra ) > sfmin2 ) {
+				f *= sclfac;
+				c *= sclfac;
+				ca *= sclfac;
+				r /= sclfac;
+				g /= sclfac;
+				ra /= sclfac;
+			}
+
+			g = c / sclfac;
+
+			while ( g >= r && max( r, ra ) < sfmax2 && min( f, c, g, ca ) > sfmin2 ) {
+				f /= sclfac;
+				c /= sclfac;
+				g /= sclfac;
+				ca /= sclfac;
+				r *= sclfac;
+				ra *= sclfac;
+			}
+
+			// Now balance
+			if ( ( c + r ) >= factor * s ) {
+				ia1 += strideA2;
+				ia2 += strideA1;
+				is += strideScale;
+				ia3 += strideA2;
+				continue;
+			}
+
+			if ( f < 1.0 && scale[ is ] < 1.0 ) {
+				if ( f * scale[ is ] <= sfmin1 ) {
```

Comment:
Can these conditionals not be consolidated?

```suggestion
			if ( f < 1.0 && scale[ is ] < 1.0 && f*scale[ is ] <= sfmin1 ) {
```

Applies here and below.